ci-lib.nix

```diff
@@ -30,16 +30,16 @@ in rec {
   # TODO: Surely this shouldn't be necessary. I think normal 'filterAttrsRecursive' will effectively cause infinite loops
   # if you keep derivations and your predicate forces the value of the attribute, as this then triggers a loop on the
   # 'out' attribute. Weird.
+  # To make this function faster, unwanted attributes are mapped to {} instead of being
+  # removed. This keeps the function lazy and avoids unwanted evaluation of sibling
+  # derivations.
   filterAttrsOnlyRecursive = pred: set:
-    lib.listToAttrs (
-      lib.concatMap (name:
-        let v = set.${name}; in
-        if pred name v then [
-          (lib.nameValuePair name (
-            if builtins.isAttrs v && !lib.isDerivation v then filterAttrsOnlyRecursive pred v
+    lib.mapAttrs (name: v:
+      if pred name v
+        then
+          if builtins.isAttrs v
+              && !lib.isDerivation v
+            then filterAttrsOnlyRecursive pred v
             else v
-          ))
-        ] else []
-      ) (builtins.attrNames set)
-    );
+        else {}) set;
 }
```

Comment:
@michaelpj I think the problem with removing attributes from the attr set is that the list of names is strict and therefor winds up depending on the evaluation of the predicate for each attribute in the set (even if you are just observing one nix wants to know which of them are still there).  mapAttrs does not replace any attributes so the list of names does not depend on anything other than the list of names in the source attribute set (no need to evaluate the predicate except on the path of the attribute we are looking for).   I was testing this by running:

Before:
```
~/iohk/haskell.nix$ time nix-build release.nix -A R1903.darwin.native.tests.cabal-simple.run
trace: Using latest index state for hello!
trace: Using index-state: 2020-04-15T00:00:00Z for hello
trace: Using latest index state for source-test-cabal-22!
trace: Using index-state: 2020-04-15T00:00:00Z for source-test-cabal-22
trace: Using latest index state for source-test-cabal-simple!
trace: Using index-state: 2020-04-15T00:00:00Z for source-test-cabal-simple
trace: Using index-state: 2019-10-01T00:00:00Z for source-test-shell-for-setup-deps
trace: Get `stack-sha256` with `nix-hash --base32 --type sha256 /nix/store/sf022w27dm8wcx11w1bb6lpcbz3y053r-source-test-stack-local-resolver-stack-to-nix-pkgs/`
trace: Get `stack-sha256` with `nix-hash --base32 --type sha256 /nix/store/rk0gcx25h4wdigrlc23piz6qx43w8wi0-source-test-stack-source-repo-stack-to-nix-pkgs/`
/nix/store/rvjlvs7ws4d2j2vf0yyv3l0483gavs8h-cabal-simple-test

real	0m5.899s
user	0m5.351s
sys	0m0.640s
```

After:
```
~/iohk/haskell.nix$ time nix-build release.nix -A R1903.darwin.native.tests.cabal-simple.run
trace: Using latest index state for source-test-cabal-simple!
trace: Using index-state: 2020-04-15T00:00:00Z for source-test-cabal-simple
/nix/store/rvjlvs7ws4d2j2vf0yyv3l0483gavs8h-cabal-simple-test

real	0m4.499s
user	0m4.279s
sys	0m0.478s
```

Reply:
Nice! I stole the implementation from nixpkgs, but I suspect it's not used much, as it's quite easy to shoot yourself in the foot with it. This seems like a sensible improvement :)